File: 리트코드/0005-longest-palindromic-substring/0005-longest-palindromic-substring.java
class Solution {
    int left;
    int maxLen;
    public String longestPalindrome(String s) {
        int len = s.length();

        if (len < 2) return s;

        for (int i = 0; i < len - 1; i++) {
            extendPalindrome(s,i, i+1);
            extendPalindrome(s,i, i+2);
        }

        return s.substring(left, left + maxLen);
    }

    public void extendPalindrome(String s, int j, int k) {
        while (0 <= j && k < s.length()&& s.charAt(j) == s.charAt(k)) {
            j--;
            k++;
        }
        if (maxLen < k - j - 1) {
            maxLen = k - j - 1;
            left = j + 1;
        }
        
    }
}

File: 리트코드/0005-longest-palindromic-substring/0005-longest-palindromic-substring.kt
class Solution {
    var left: Int = 0;
    var maxLen: Int = 0;

    fun longestPalindrome(s: String): String {
        val len = s.length

        if(len < 2) return s
        
        for (i in 0 until  len - 1) {
            expendPalindrome(s,i,i+1)
            expendPalindrome(s,i,i+2)
        }
        return s.substring(left,left + maxLen)
    }
    fun expendPalindrome(s: String, j: Int, k: Int) {
        var l = j
        var r = k
        while (0 <= l && r < s.length && s[l] == s[r]) {
            l--
            r++
        }

        if (maxLen < r - l - 1) {
            maxLen = r - l - 1
            left = l + 1
        }
    }
}